### PR TITLE
Put Local, Static and Promoted as one Base variant of Place

### PR DESCRIPTION
**Pull request by spastorino**
Related to #52708

The `Place` 2.0 representation use a `Base` variant for `Local`, `Static` and `Promoted` so we start making this change in the current `Place` to make the following steps simpler.

r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 1]
```diff
@@ -1813,7 +1815,9 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
     fn classify_drop_access_kind(&self, place: &Place<'tcx>) -> StorageDeadOrDrop<'tcx> {
         let tcx = self.infcx.tcx;
         match place {
-            Place::Local(_) | Place::Static(_) | Place::Promoted(_) => {
+            Place::Base(PlaceBase::Local(_)) |
+                Place::Base(PlaceBase::Static(_)) |
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/error_reporting.rs]: the indent is off here

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 2]
```diff
@@ -1910,7 +1915,10 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                     "annotate_argument_and_return_for_borrow: target={:?} stmt={:?}",
                     target, stmt
                 );
-                if let StatementKind::Assign(Place::Local(assigned_to), box rvalue) = &stmt.kind
+                if let StatementKind::Assign(
+                            Place::Base(PlaceBase::Local(assigned_to)),
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/error_reporting.rs]: odd indentation here. Maybe write

```rust
if let StatementKind::Assign(
    Place::Base(PlaceBase::Local(assigned_to)),
    box rvalue,
) = &stmt.kind {
```

File: src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs [hunk 3]
```diff
@@ -384,7 +384,8 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                         // Just point to the function, to reduce the chance of overlapping spans.
                         let function_span = match func {
                             Operand::Constant(c) => c.span,
-                            Operand::Copy(Place::Local(l)) | Operand::Move(Place::Local(l)) => {
+                            Operand::Copy(Place::Base(PlaceBase::Local(l))) |
+                                Operand::Move(Place::Base(PlaceBase::Local(l))) => {
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs]: use the same indent as the `Operand::Copy`

File: src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs [hunk 4]
```diff
@@ -462,17 +463,19 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                         // If we see a use, we should check whether it is our data, and if so
                         // update the place that we're looking for to that new place.
                         Rvalue::Use(operand) => match operand {
-                            Operand::Copy(Place::Local(from)) |
-                            Operand::Move(Place::Local(from)) if *from == target => {
+                            Operand::Copy(Place::Base(PlaceBase::Local(from))) |
+                                Operand::Move(Place::Base(PlaceBase::Local(from)))
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs]: indent

File: src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs [hunk 5]
```diff
@@ -462,17 +463,19 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
                             },
                             _ => {},
                         },
                         // If we see a unsized cast, then if it is our data we should check
                         // whether it is being cast to a trait object.
                         Rvalue::Cast(CastKind::Unsize, operand, ty) => match operand {
-                            Operand::Copy(Place::Local(from)) |
-                            Operand::Move(Place::Local(from)) if *from == target => {
+                            Operand::Copy(Place::Base(PlaceBase::Local(from))) |
+                                Operand::Move(Place::Base(PlaceBase::Local(from)))
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs]: indent

File: src/librustc_mir/borrow_check/prefixes.rs [hunk 6]
```diff
@@ -26,8 +26,9 @@ impl<'tcx> IsPrefixOf<'tcx> for Place<'tcx> {
             }
 
             match *cursor {
-                Place::Promoted(_) |
-                Place::Local(_) | Place::Static(_) => return false,
+                Place::Base(PlaceBase::Promoted(_)) |
+                Place::Base(PlaceBase::Local(_)) |
+                Place::Base(PlaceBase::Static(_)) => return false,
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/prefixes.rs]: Just use a single pattern (`Place::Base(_)`)

File: src/librustc_mir/borrow_check/prefixes.rs [hunk 7]
```diff
@@ -86,9 +87,9 @@ impl<'cx, 'gcx, 'tcx> Iterator for Prefixes<'cx, 'gcx, 'tcx> {
 
         'cursor: loop {
             let proj = match *cursor {
-                Place::Promoted(_) |
-                Place::Local(_) | // search yielded this leaf
-                Place::Static(_) => {
+                Place::Base(PlaceBase::Promoted(_)) |
+                Place::Base(PlaceBase::Local(_)) | // search yielded this leaf
+                Place::Base(PlaceBase::Static(_)) => {
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/prefixes.rs]: same here, just use the `Place::Base(_)` pattern

File: src/librustc_mir/build/expr/as_rvalue.rs [hunk 8]
```diff
@@ -127,19 +127,24 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
...
                 }
 
                 // malloc some memory of suitable type (thus far, uninitialized):
                 let box_ = Rvalue::NullaryOp(NullOp::Box, value.ty);
                 this.cfg
-                    .push_assign(block, source_info, &Place::Local(result), box_);
+                    .push_assign(block, source_info, &Place::Base(PlaceBase::Local(result)), box_);
 
                 // initialize the box contents:
-                unpack!(block = this.into(&Place::Local(result).deref(), block, value));
```
Comment (oli-obk) [inline src/librustc_mir/build/expr/as_rvalue.rs]: pull `Place::Local(result)` into a variable before this statement and use that variable here and in the next statement

File: src/librustc_mir/build/mod.rs [hunk 9]
```diff
@@ -956,7 +956,7 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
         }
 
         let body = self.hir.mirror(ast_body);
-        self.into(&Place::Local(RETURN_PLACE), block, body)
+        self.into(&Place::Base(PlaceBase::Local(RETURN_PLACE)), block, body)
```
Comment (oli-obk) [inline src/librustc_mir/build/mod.rs]: Maybe add a `Place::RETURN_PLACE` associated constant to `Place` to make all of these `Place::Base(PlaceBase::Local(RETURN_PLACE))` shorter.

File: src/librustc_mir/transform/add_retag.rs [hunk 10]
```diff
@@ -21,9 +21,9 @@ fn is_stable<'tcx>(
 
     match *place {
         // Locals and statics have stable addresses, for sure
-        Local { .. } |
-        Promoted { .. } |
-        Static { .. } =>
+        Base(PlaceBase::Local { .. }) |
```
Comment (oli-obk) [inline src/librustc_mir/transform/add_retag.rs]: Just use `Base(_)`

File: src/librustc_mir/util/elaborate_drops.rs [hunk 11]
```diff
@@ -703,9 +703,9 @@ impl<'l, 'b, 'tcx, D> DropCtxt<'l, 'b, 'tcx, D>
         };
 
         let cur = self.new_temp(iter_ty);
-        let length = Place::Local(self.new_temp(tcx.types.usize));
+        let length = Place::Base(PlaceBase::Local(self.new_temp(tcx.types.usize)));
         let length_or_end = if ptr_based {
-            Place::Local(self.new_temp(iter_ty))
+            Place::Base(PlaceBase::Local(self.new_temp(iter_ty)))
```
Comment (oli-obk) [inline src/librustc_mir/util/elaborate_drops.rs]: Add a FIXME to `new_temp` to check if we want to make it return a `Place` directly if all use sites want a `Place::Base` anyway.

**Review by oli-obk — CHANGES_REQUESTED** [review]
lgtm

Note that this will cause a max-rss regression, so we'll merge it after the beta cutoff, so we have 6 weeks to fix it (which should happen with the `enum Place` -> `struct Place` refactoring discussed in https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/place.202.2E0/near/159082516

File: src/librustc_mir/borrow_check/error_reporting.rs (hunk 1 above)
Comment (spastorino) [reply src/librustc_mir/borrow_check/error_reporting.rs]: Yeah, vim does that :). Fixing it.

File: src/librustc_mir/borrow_check/prefixes.rs (hunk 6 above)
Comment (spastorino) [reply src/librustc_mir/borrow_check/prefixes.rs]: Hmm kind of disagree with this one. What if we add something in `Base` that needs a different treatment. For instance the All Hands docs talks about adding `Deref` here. Unsure if that will need to be treated differently.

File: src/librustc_mir/borrow_check/prefixes.rs (hunk 7 above)
Comment (spastorino) [reply src/librustc_mir/borrow_check/prefixes.rs]: Same as above.

File: src/librustc_mir/borrow_check/prefixes.rs (hunk 6 above)
Comment (oli-obk) [reply src/librustc_mir/borrow_check/prefixes.rs]: Sure, fully listing all variants is a good regression check.

File: src/librustc_mir/build/expr/as_rvalue.rs (hunk 8 above)
Comment (spastorino) [reply src/librustc_mir/build/expr/as_rvalue.rs]: `deref` takes ownership so that's not possible

**Comment by spastorino** [comment]
@oli-obk fixed, rebased and force pushed :).

**Review by oli-obk — APPROVED** [review]

**Comment by matthewjasper** [comment]
>  so we'll merge it after the beta cutoff

@oli-obk What happened to this?

**Comment by spastorino** [comment]
@oli-obk just in case conversation in Zulip went unnoticed. This is ready to r+.

**Comment by spastorino** [comment]
@oli-obk @Centril There were conflicts, just force pushed again.

**Comment by Centril** [comment]
Seems to be a bit-rotty PR, so @bors p=20

**Comment by nnethercote** [comment]
FWIW, I was just looking at peak memory usage and I identified the `Place`/`PlaceBase` split as a source of wasted space and I started combining them and then I thought "hmm..." and I used `git blame` to find this PR.

In other words, I noticed this change.

**Comment by spastorino** [comment]
@nnethercote this was an intermediate step, we are going after this ...

```rust
struct Place {
    base: PlaceBase,
    projection: PlaceProjection,
}
enum PlaceProjection {
    Base,
    Projection(Box<PlaceProjection>),
    Deref,
    Index(..),
    ...
}
```

Going to provide a PR for this probably this week.

**Comment by matthewjasper** [comment]
That's makes Place even larger.

**Comment by oli-obk** [comment]
Since the final layout will be

```rust
struct Place<'tcx> {
    base: PlaceBase,
    projection: &'tcx [PlaceProjection],
}
enum PlaceProjection {
    Projection(Box<PlaceProjection>),
    Deref,
    Index(..),
    ...
}
```

we can also box `Place` at all/most use sites, so any uses of it will just be one pointer sized.

I think that ideally we'd be able to have all occurences of `Place` be `&'tcx Place` and the type be

```rust
struct Place {
    base: PlaceBase,
    projection: [PlaceProjection],
}
```

But I don't know how to initialize such a type without statically knowing the length of the `projection` field.

**Comment by eddyb** [comment]
FWIW `&'tcx [PlaceProjection]` is *not* an interned projection list.
Look at `ty::List`, which is how the interner keeps things, and note that the length is *behind* the pointer.

Also, we shouldn't combine the base and the projections, IMO.
Ideally the base can change without the projections needing reinterning (think renaming locals).

**Comment by eddyb** [comment]
I expect `PlaceBase` to be 4 bytes in size eventually (locals- anything other than a local should always be accessed via a borrow, which can allow e.g. borrowing statics/promoteds - by-value access for those doesn't really make sense anyway)

And a `ty::List<PlaceProjection>` pointer is thin, so `Place` would be twice the pointer width.

@nnethercote Does that work for you?

**Comment by spastorino** [comment]
> That's makes `Place` even larger.

Yes, sorry. I've pasted an intermediate step. The final layout as @eddyb has said is ...

```rust
struct Place<'tcx> {
    base: PlaceBase,
    projection: &'tcx [PlaceProjection],
}
enum PlaceProjection {
    Projection(Box<PlaceProjection>),
    Deref,
    Index(..),
    ...
}
```

**Comment by oli-obk** [comment]
Not quite, @eddyb is right, `&'tcx [T]` is a pointer and a length, while `&'tcx ty::List<T>` is one pointer, so we'll want to have

```rust
struct Place<'tcx> {
    base: PlaceBase,
    projection: &'tcx ty::List<PlaceProjection>,
}
enum PlaceProjection {
    Projection(Box<PlaceProjection>),
    Deref,
    Index(..),
    ...
}
```

**Comment by RalfJung** [comment]
What is the point of the recursive `Projection(Box<PlaceProjection>)` variant?

**Comment by oli-obk** [comment]
Ups, that's just a copy paste bug. In the `List` situation we won't have any more recursion in the future, that's just a leftover right now. The next step is actually

```rust
enum Place {
    Base(PlaceBase)
    Projection(Box<PlaceProjection>),
}
struct PlaceProjection {
    base: Place,
    projection: PlaceProjectionElem,
}
enum PlaceProjectionElem {
    Leaf,
    Deref,
    Index(..),
    ...
}
```

while the target situation is

```rust
struct Place {
    base: PlaceBase
    projections: &'tcx ty::List<PlaceProjectionElem>,
}
// Note: no more PlaceProjection
enum PlaceProjectionElem {
    // Note: no more "Leaf"
    Deref,
    Index(..),
    ...
}
```

**Comment by nnethercote** [comment]
> @nnethercote Does that work for you?

I'm a bit confused by all the above back and forth. `Place` was 16 bytes before this started, which was good. Then it became 24 bytes, which is worse. 32 bytes would be even worse. Smaller is better.

**Comment by eddyb** [comment]
@nnethercote The plan is for it to become 16 bytes (`PlaceBase` the size of `Local`, plus a `&'tcx ty::List<PlaceProjectionElem>` which is pointer-sized).
But it will get worse before it gets better, otherwise we'll never get where we want to get.

But also, `Place` being small is not *by itself* good. It was/is a linked list, which is much worse than an array.

**Comment by spastorino** [comment]
Yeah, sorry my bad I was confusing you a lot. Thanks @eddyb for clarying it.
The last version as stated will be 24 bytes.

**Comment by nnethercote** [comment]
> But also, `Place` being small is not by itself good.

I can see that an array is better than a linked list. But note that the size of `Place` affects the size of `mir::Statement` So even statements that don't contain a `Place` field get bigger if `Place` gets bigger. And the size of `mir::Statement` affects peak memory size.

Anyway, if the final size will be 16 bytes, that's great.

**Pull request merged**